### PR TITLE
Improve the error message for non-associative comparisons.

### PR DESCRIPTION
**Pull request by DanilaFe**
This PR builds on https://github.com/chapel-lang/chapel/pull/24155. Closes https://github.com/chapel-lang/chapel/issues/24559.

In https://github.com/chapel-lang/chapel/pull/24155, we made `a < b < c` a syntax error, because it leads to confusing results: a boolean is implicitly converted to a numeric value, and compared against another number. I went about disallowing such syntax by simply removing the associativity from `<`, `>`, etc. However, this meant that Bison simply reported "syntax error" when it encountered such an expression, which leaves much to be desired.

This PR improves on that situation by making adjustments to the parser. Specifically, it enables the parser to note whether or not parentheses were used for an expression (including an operator comparison). This is done using the additional location maps added by @dlongnecke-cray in https://github.com/chapel-lang/chapel/pull/23548. The parser is thus allowed to accept `a < b < c`, and subsequent checks are executed that detect `a < b < c` (as opposed to `(a < b) < c`, which is valid). These checks use the Dyno error system to print a nice error message.

Screen Shot 2024-03-07 at 3 45 34 PM

Reviewed by @dlongnecke-cray -- thanks!

## Testing
- [x] paratest
- [x] dyno tests

### CODE REVIEW

File: frontend/lib/uast/post-parse-checks.cpp [hunk 1]
```diff
@@ -709,6 +711,59 @@ void Visitor::checkDmappedKeyword(const OpCall* node) {
            " instead please use factory functions when available");
 }
 
+static int getOpCallPrecedence(const OpCall* node) {
```
Comment (dlongnecke-cray) [inline frontend/lib/uast/post-parse-checks.cpp:714]: Maybe drop this since it's a one liner.

File: frontend/lib/uast/post-parse-checks.cpp [hunk 2]
```diff
@@ -709,6 +711,59 @@ void Visitor::checkDmappedKeyword(const OpCall* node) {
...
+
+  check(node->actual(0));
+  ops.push_back(node);
+  check(node->actual(1));
+}
+
+void Visitor::checkNonAssociativeComparisons(const OpCall* node) {
+  if (node->numActuals() != 2) return;
+
+  auto lessThanPrec = opToPrecedence(USTR("<"), /* unary */ false, /* postfix */ false);
```
Comment (dlongnecke-cray) [inline frontend/lib/uast/post-parse-checks.cpp:745]: Consider just `const bool unary = false;` and `const bool postfix = false;`.

File: frontend/lib/parsing/parser-error-classes-list.cpp [hunk 3]
```diff
@@ -172,6 +172,76 @@ void ErrorMultipleExternalRenaming::write(ErrorWriterBase& wr) const {
...
+  // Bitfield: 001 for < or <=, 010 for > or >=, 100 for == or !=
+  // We only want to suggest the "normalized" form if all operators are
+  // in the same direction, and if no equality constraints are present.
+  //
+  // The reason for the latter is, how would you desugar x != y != z?
+  //     x != y && y != z
+  // or
+  //     x != y && y != z && x != z
+  // It's best not to guess there.
+  int types = 0;
```
Comment (dlongnecke-cray) [inline frontend/lib/parsing/parser-error-classes-list.cpp:193]: Clever.

**Review by dlongnecke-cray — APPROVED** [review]
Nice. Just silly nitpicks.

File: frontend/lib/uast/post-parse-checks.cpp (hunk 2 above)
Comment (DanilaFe) [reply frontend/lib/uast/post-parse-checks.cpp:745]: I've combined this into 

```
static int binOpPrecedence(UniqueString ustr) {
  bool unary = false;
  bool postfix = false;
  return opToPrecedence(ustr, unary, postfix);
}
```

Which makes it no longer a one-liner, and saves some repetition.

**Pull request merged**